Switch to the Chronos HTTP client (adds support for HTTPS)

zah opened the pull request.

cheatfate commented on json_rpc/clients/httpclient.nim:
```diff
@@ -205,22 +64,28 @@ method call*(client: RpcHttpClient, name: string,
+  let addresses = resolveTAddress(address, port)
+  if addresses.len == 0:
+    raise newException(RpcAddressUnresolvableError, "Failed to resolve address: " & address)
+  ok client.httpAddress, getAddress(addresses[0])
```
Using this approach you will going to stick with the only one IP address you have resolved. But chronos HTTP client could perform resolve and use all the IP addresses. So its better here to use https://github.com/status-im/nim-chronos/blob/master/chronos/apps/http/httpclient.nim#L366-L367 this `getAddress`.

I think you can easily convert `address` and `port` to `http://<address>:<port>`.

cheatfate replied:
But if you suppose that `address` is an IP address here, you should use `initTAddress` which will return single `TransportAddress` instance, which could be used to create `HttpAddress`.

bernardoaraujor replied:
https://github.com/bernardoaraujor/nim-json-rpc/commit/7e1846c30110b5bf787ce5be6bd484453ed9afd4

cheatfate commented:
I dont see HTTPS support here.
Which type of connection is going to be established depends on `HttpAddress`. And right now i see only ability to create non-secure connections. All addresses should be converted to URL and `getAddress(URL)` should be used to obtain `HttpAddress` information.

cheatfate commented on json_rpc/clients/httpclient.nim:
```diff
@@ -14,188 +16,45 @@ type
+    req = HttpClientRequestRef.post(client.httpSession,
+                                    client.httpAddress.get,
+                                    body = reqBody.toOpenArrayByte(0, reqBody.len - 1))
+    res = await req.send()
```
This could raise `HttpError`.

bernardoaraujor replied:
https://github.com/bernardoaraujor/nim-json-rpc/commit/98d79570f83f743c756177e7f0b49cdc56c7a2c3

cheatfate commented on json_rpc/clients/httpclient.nim:
```diff
@@ -14,188 +16,45 @@ type
-  await transp.closeWait()
-  if value.len == 0:
-    raise newException(ValueError, "Empty response from server")
+  let resText = string.fromBytes(await res.getBodyBytes(client.maxBodySize))
```
This could raise `HttpError`.

bernardoaraujor replied:
@cheatfate I'm checking [the definition of `getBodyBytes`](https://github.com/status-im/nim-chronos/blob/14ebf269e9322de5a7f1fa455033b0bcf18144c6/chronos/apps/http/httpclient.nim#L1100) and couldn't find `HttpError` being raised.
Perhaps you meant `CancelledError` and `AsuncStreamError`?

bernardoaraujor replied:
https://github.com/bernardoaraujor/nim-json-rpc/commit/6a52c42ea0d95540fc1d74125c4afd67b824052d

cheatfate commented on json_rpc/clients/httpclient.nim:
```diff
@@ -14,188 +16,45 @@ type
-    raise newException(ValueError, "Empty response from server")
+  let resText = string.fromBytes(await res.getBodyBytes(client.maxBodySize))
+  trace "Response", text = resText
+  echo "response ", resText
```
I think this `echo` should not be present here.

bernardoaraujor replied:
does that also apply to [` echo "req body ", reqBody`](https://github.com/status-im/nim-json-rpc/blob/364dd2ebd132d7eb5e4516099250060b094c3daa/json_rpc/clients/httpclient.nim#L53)?

bernardoaraujor replied:
https://github.com/bernardoaraujor/nim-json-rpc/commit/0b70c35a380d51d94a12e079e9f7316481e7a987

zah commented:
@bernardoaraujor is planning to work on addressing the review comments in this branch and on adding a matching implementation of a HTTP server based on Chronos.

bernardoaraujor commented:
> @bernardoaraujor is planning to work on addressing the review comments in this branch and on adding a matching implementation of a HTTP server based on Chronos.

@zah how should I submit work here? create a new PR from a fork into `chronos-http`?

bernardoaraujor commented:
@zah another question:

I'm exploring `tests/testhttp.nim` as a starting point for the Chronos HTTP replacement.
The test suite is currently breaking with:
```
/home/bear/.choosenim/toolchains/nim-1.4.8/lib/pure/unittest.nim(654) testhttp
/home/bear/.nimble/pkgs/chronos-3.0.5/chronos/asyncloop.nim(1091) waitFor
/home/bear/.nimble/pkgs/chronos-3.0.5/chronos/asyncloop.nim(279) poll
/home/bear/develop/nim-json-rpc/json_rpc/clients/httpclient.nim(67) call
/home/bear/develop/nim-json-rpc/json_rpc/client.nim(44) processMessage
/home/bear/.choosenim/toolchains/nim-1.4.8/lib/pure/json.nim(991) parseJson
/home/bear/.choosenim/toolchains/nim-1.4.8/lib/pure/json.nim(902) parseJson
/home/bear/.choosenim/toolchains/nim-1.4.8/lib/pure/json.nim(867) parseJson
/home/bear/.choosenim/toolchains/nim-1.4.8/lib/pure/parsejson.nim(522) raiseParseErr
[[reraised from:
/home/bear/.choosenim/toolchains/nim-1.4.8/lib/pure/unittest.nim(654) testhttp
/home/bear/.nimble/pkgs/chronos-3.0.5/chronos/asyncloop.nim(1091) waitFor
/home/bear/.nimble/pkgs/chronos-3.0.5/chronos/asyncloop.nim(279) poll
/home/bear/develop/nim-json-rpc/json_rpc/clients/httpclient.nim(67) call
/home/bear/develop/nim-json-rpc/json_rpc/client.nim(45) processMessage
]]
[[reraised from:
/home/bear/.choosenim/toolchains/nim-1.4.8/lib/pure/unittest.nim(654) testhttp
/home/bear/.nimble/pkgs/chronos-3.0.5/chronos/asyncloop.nim(1091) waitFor
/home/bear/.nimble/pkgs/chronos-3.0.5/chronos/asyncloop.nim(279) poll
/home/bear/.nimble/pkgs/chronos-3.0.5/chronos/asyncmacro2.nim(371) continuousTest
/home/bear/.nimble/pkgs/chronos-3.0.5/chronos/asyncfutures2.nim(503) internalCheckComplete
]]
[[reraised from:
/home/bear/.choosenim/toolchains/nim-1.4.8/lib/pure/unittest.nim(654) testhttp
/home/bear/.nimble/pkgs/chronos-3.0.5/chronos/asyncloop.nim(1093) waitFor
/home/bear/.nimble/pkgs/chronos-3.0.5/chronos/asyncfutures2.nim(517) read
/home/bear/.nimble/pkgs/chronos-3.0.5/chronos/asyncfutures2.nim(503) internalCheckComplete
]]

    Unhandled exception: input(1, 0) Error: { expected [JsonParsingError]
```

Upon a closer examination, I arrive [here](https://github.com/status-im/nim-json-rpc/blob/364dd2ebd132d7eb5e4516099250060b094c3daa/json_rpc/client.nim#L46), so the [issue](https://github.com/status-im/nimbus-eth2/issues/2430) seems to be well, known.

If I comment out `continuousTest`:
```
 suite "HTTP Server/HTTP Client RPC test suite":
-  test "Continuous RPC calls (" & $TestsCount & " messages)":
-    check waitFor(continuousTest("localhost", Port(8545))) == TestsCount
+  # test "Continuous RPC calls (" & $TestsCount & " messages)":
+  #  check waitFor(continuousTest("localhost", Port(8545))) == TestsCount
```

... the suite runs fine.

I don't intend to commit the removal of `continuousTest`, but for the moment I'm going to proceed with the Chronos HTTP replacement while simply ignoring the Continuous RPC calls tests, is that ok?

jangko commented:
Because you need to modify the server heavily, it's ok to open a new PR. you are free to choose the merge target depend on how much the new client share code with client from this PR. If it only share insignificant amount of code, better set the merge target directly to master branch.

bernardoaraujor commented:
I made some significant progress for the usage of Chronos HTTP on the JSON-RPC module.
I've addressed the review comments from @cheatfate, fixed a few bugs, and started the Server side replacement.

The work is being commited in a [fork](https://github.com/bernardoaraujor/nim-json-rpc/) over the `chronos-http-bernardo` branch, which I will use as a base to create a PR to `chronos-http` on this repo.

Unfortunately, I'm stuck in the [Chronos HTTP Server replacement](https://github.com/bernardoaraujor/nim-json-rpc/blob/chronos-http-bernardo/json_rpc/servers/httpserver.nim) and I need some help.

The first step was for the Server replacement was:
```
RpcHttpServer* = ref object of RpcServer
-    servers: seq[StreamServer]
+    servers: seq[HttpServerRef]
```

Then I replaced the `addStreamServer` functions with [`addHttpServer` counterparts](https://github.com/bernardoaraujor/nim-json-rpc/blob/b0ecbd3e5aea9b82c6344b2df61627441e7e39e7/json_rpc/servers/httpserver.nim#L224).

The next logical step is to rewrite the `processClient` callback, which will invoke the appropriate `rpc.route` calls after each request. Note that [`HttpServerRef.new` imposes that `processClient` must be `HttpProcessCallback`](https://github.com/status-im/nim-chronos/blob/14ebf269e9322de5a7f1fa455033b0bcf18144c6/chronos/apps/http/httpserver.nim#L179).

Originally, the `RpcHttpServer` object was retrieved inside the scope of `processClient` via [`getUserData[RpcHttpServer](server)`](https://github.com/status-im/nim-json-rpc/blob/364dd2ebd132d7eb5e4516099250060b094c3daa/json_rpc/servers/httpserver.nim#L89), which was possible because [the object was stored inside the `StreamServer` as `udata`](https://github.com/status-im/nim-json-rpc/blob/364dd2ebd132d7eb5e4516099250060b094c3daa/json_rpc/servers/httpserver.nim#L215).

Since we changed the signature of `processClient`, `server` is no longer available in the scope, and `rpc` can't be accessed from its `udata`, which makes it impossible to route the RPC calls.

The first solution I could imagine would be to add a `udata*: pointer` field to [`HttpServer` ](https://github.com/status-im/nim-chronos/blob/14ebf269e9322de5a7f1fa455033b0bcf18144c6/chronos/apps/http/httpserver.nim#L63) in the [same style as `SocketServer`](https://github.com/status-im/nim-chronos/blob/14ebf269e9322de5a7f1fa455033b0bcf18144c6/chronos/transports/common.nim#L69), and adjust `HttpServerRef.new` so it optionally takes `udata` as a parameter.
But that would imply modifying `nim-chronos`, so I'm not sure that's a valid path.

Perhaps there's some other way to retrieve `RpcHttpServer` from inside the `processClient` scope?
Or am I completely misguided in my approach?

jangko commented:
you can see this passing via closure for inspiration:

https://github.com/status-im/nim-graphql/blob/5ed7fc0af123add32e1c88439c73b473fe96c46e/graphql/httpserver.nim#L188
